airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/AirbyteTracedException.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.base;
+
+import io.airbyte.protocol.models.AirbyteErrorTraceMessage;
+import io.airbyte.protocol.models.AirbyteErrorTraceMessage.FailureType;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.AirbyteMessage.Type;
+import io.airbyte.protocol.models.AirbyteTraceMessage;
+import java.util.Arrays;
+import java.util.function.Consumer;
+
+public class AirbyteTracedException extends Exception {
+
+  private final Throwable e;
+  private final String displayMessage;
+  private final FailureType failureType;
+
+  public AirbyteTracedException(final Throwable e, final String displayMessage, final FailureType failureType) {
+    this.e = e;
+    this.displayMessage = displayMessage;
+    this.failureType = failureType;
+
+    // Not sure why defaultOutputRecordCollector is under Destination specifically,
+    // but this matches IntegrationRunner usage
+    Consumer<AirbyteMessage> outputRecordCollector = Destination::defaultOutputRecordCollector;
+    outputRecordCollector.accept(new AirbyteMessage().withType(Type.TRACE).withTrace(makeAirbyteTraceMessage()));
```

Comment:
👍 creating the new class has the side-effect of emitting the log

Reply:
as a consumer of this class, I feel like I'd be very surprised by this. Why is it being done as a side effect instead of leaving that responsibility to the calling context? It would also mean we don't need to create an output record collector here, that's the calling context's responsibility

Reply:
I intended for this to be used like `throw new AirbyteTracedException(e, "here's a user friendly message");` for a very easy implementation on the consumer side, and therefore decided to go with the log emission as a side effect of creation, as opposed to 1. instantiate, 2. log trace message, 3. throw.